### PR TITLE
general tidy and bug fixes

### PR DESCRIPTION
**Pull request by reube**
Details: see comments on pull request, where things are either as bugs or refactoring/cleaning.  thx :)
Joost et al: if you verify using your scenario testing, that would be much appreciated.
I have produced unit tests for the refactored and some other code based on my understanding of what the code should do.  
I have suggested fixes for all of the bugs identified, except for one in `position.py`.  Pls see comments in `test_position.py`.
Also, please expand `route.py`.  As I mainly focused my efforts here, the diff does not automatically render! :)

### CODE REVIEW

**Comment by ProfHoekstra** [comment]
Hi Andrew, 

If you adopt the few findact() changes in route.py from the master (units & sqrt in dist calculation, 1 deg lat = 60 nm), I think this should resolve the conflict in route.py.

Best regards,
    Jacco

**Comment by jooste** [comment]
Hi Andrew,

I've had a first look through the pull request. Really nice that you're working on unit tests! A couple of questions so far:

- You assert a lot of float comparisons with ==. From what I understand this is tricky business, and it is better to check whether a float is within a certain bound (e.g., assert abs(result - reference) < threshold).

- You've added `pandir` as a possible content for the `Position` object. What is your idea behind this? We currently only have `pandir` for parsing PAN stack commands; they aren't used anywhere else.

File: bluesky/traffic/route.py [hunk 1]
```diff
@@ -79,147 +108,130 @@ def addwptStack(self, idx, *args):  # args: all arguments of addwpt
...
+                afterwp = self.wpname[rwyrte_idx]
 
-            elif self.wptype and self.wptype[0] == self.orig:
+            elif self.wptype and self.wptype[0] == Route.orig:
                 afterwp = self.wpname[0]
 
-            else:
-                # Assume we're called before other waypoints are added
-                afterwp = ""
-
```
Comment (reube) [inline bluesky/traffic/route.py]: This else is redundant as the initializer sets it.

File: bluesky/traffic/route.py [hunk 2]
```diff
@@ -79,147 +108,130 @@ def addwptStack(self, idx, *args):  # args: all arguments of addwpt
...
-            name = "T/O-" + bs.traf.id[idx] # Use lat/lon naming convention
+            name = "T/O-" + bs.traf.id[idx]  # Use lat/lon naming convention
         # Add waypoint
-        wpidx = self.addwpt(idx, name, wptype, lat, lon, alt, spd, afterwp, beforewp)
+        wpok, wpidx = self.addwpt(
+            idx, name, wptype, lat, lon, alt, spd, afterwp, beforewp)
```
Comment (reube) [inline bluesky/traffic/route.py]: Now addwpt returns ok status and the index

File: bluesky/traffic/route.py [hunk 3]
```diff
@@ -253,15 +270,227 @@ def afteraddwptStack(self, idx, *args):  # args: all arguments of addwpt
...
+            self.wpname.insert(wpidx, wpname_)
+            self.wplat.insert(wpidx, wplat_)
+            self.wplon.insert(wpidx, wplon_)
+            self.wpalt.insert(wpidx, wpalt_)
+            self.wpspd.insert(wpidx, wpspd_)
+            self.wptype.insert(wpidx, wptype_)
+            self.wpflyby.insert(wpidx, swflyby_)
+
+    def addwpt_orig_dest(self, name_, wprtename, wptype, lat, lon, alt, spd):
+        """
```
Comment (reube) [inline bluesky/traffic/route.py]: This is essentially the same code as before - mainly it's been just refactored out of `addwpt`.  
As there is fair commonality between adding of origin and destination waypoints, I have combined these actions into a common function, where variables such as `wpidx` are appropriately set based on type (i.e. to 0 for Route.orig and -1 for Route.dest).  

A few things have been tweaked such as incrementing `nwp` only if inserting a wp, not overwriting.

File: bluesky/traffic/route.py [hunk 4]
```diff
@@ -253,15 +270,227 @@ def afteraddwptStack(self, idx, *args):  # args: all arguments of addwpt
...
+                self.iactwp += 1
+            elif not orig and self.iactwp < 0 and self.nwp == 1:
+                # When only waypoint: adjust pointer to point to destination
+                self.iactwp = 0
+
+        return True, 0 if orig else self.nwp - 1
+
+    def addwpt_normal(self, idx, name_, wprtename, wptype, lat, lon,
+                      alt, spd, afterwp, beforewp):
+        """
```
Comment (reube) [inline bluesky/traffic/route.py]: Similarly, new function for 'normal' waypoints.

File: bluesky/traffic/route.py [hunk 5]
```diff
@@ -274,19 +503,18 @@ def atwptStack(self, idx, *args):  # args: all arguments of addwpt
                 # acid AT wpinroute SPD: show spd constraint at this waypoint
                 # acid AT wpinroute ALT: show alt constraint at this waypoint
 
-                txt = name + " : "
-
```
Comment (reube) [inline bluesky/traffic/route.py]: line 277 in original is redundant `txt = name + " : "`

File: bluesky/traffic/route.py [hunk 6]
```diff
@@ -274,19 +503,18 @@ def atwptStack(self, idx, *args):  # args: all arguments of addwpt
...
                 # acid AT wpinroute ALT: show alt constraint at this waypoint
 
-                txt = name + " : "
-
-                if len(args)==1 or (len(args)==2 and not args[1].count("/")==1):
+                if len(args) == 1 or \
+                        (len(args) == 2 and not args[1].count("/") == 1):
 
-                    txt = name+" : "
+                    txt = name + " : "
```
Comment (reube) [inline bluesky/traffic/route.py]: as its set immediately afterwards

File: bluesky/traffic/route.py [hunk 7]
```diff
@@ -296,44 +524,44 @@ def atwptStack(self, idx, *args):  # args: all arguments of addwpt
                     # Show altitude
                     if swalt:
                         if self.wpalt[wpidx] < 0:
-                            txt = txt+"-----"
+                            txt += "-----"
 
                         elif self.wpalt[wpidx] > 4500 * ft:
-                            FL = int(round((self.wpalt[wpidx]/(100.*ft))))
-                            txt = txt+"FL"+str(FL)
+                            fl = int(round((self.wpalt[wpidx] / (100. * ft))))
```
Comment (reube) [inline bluesky/traffic/route.py]: function vars in python should be lower case.

File: bluesky/traffic/route.py [hunk 8]
```diff
@@ -886,152 +847,145 @@ def newcalcfp(self):
...
                 x.append(dist2go[i])
-                name.append(self.wpname[i]+" ")    # space for check first 2 chars later
+                name.append(self.wpname[i] + " ")
+                # space for check first 2 chars later
 
         # Find where to insert cruise segment (if any)
 
         # Find longest segment without altitude constraints
 
+        desslope = clbslope = 1.
```
Comment (reube) [inline bluesky/traffic/route.py]: code analyzer complains about possible non-initialization of `desslope` and `clbslope`

**Comment by reube** [comment]
Hi both,
The float assertions are ropey at the moment, and that is something I need to improve.  I have been mainly converting them to ints and doing an assertion based on that, which i guess is similar to the thresholding you give above.  I'll improve these over the weekend - no hurry on the pull request.

Re PAN, as part of writing regression tests for the Position class, I need to cover this path in the original code:

```
 elif name.upper() in ["LEFT","RIGHT","ABOVE","DOWN"]:
            self.lat = reflat
            self.lon = reflon
            self.type = "dir" 
```
So, the type is set to `dir` based on `name` being one of the four given.  

(tests in: https://github.com/ethertricity/bluesky/blob/2ae8075a82fffa4650a2ff28b008aae04adf976f/bluesky/test/traffic/test_position.py)

The unit test fails because `ABOVE` appears to be also the name of a waypoint, and thus the code succeeds here instead:

```
# fix or navaid?
elif bs.navdb.wpid.count(name) > 0:
    idx = bs.navdb.getwpidx(name, reflat, reflon)
    self.lat = bs.navdb.wplat[idx]
    self.lon = bs.navdb.wplon[idx]
    self.type = Position.nav
```

and never gets parsed as a PAN dir.   Perhaps, `UP` could be used instead of `ABOVE`, or something similar?

**Comment by reube** [comment]
floats assertions are now handled more gracefully :)

**Pull request closed without merging**